### PR TITLE
Fjmalass nodes as tuples

### PR DESCRIPTION
**Pull request by fjmalass**
fix for #1975 

Creation of to_tuple in "utils.misc" and converted nodes to tuples in node_linnk_graph.

### CODE REVIEW

File: networkx/readwrite/json_graph/tests/test_node_link.py [hunk 1]
```diff
@@ -44,6 +47,21 @@ def test_multigraph(self):
         nx.is_isomorphic(G, H)
         assert_equal(H[1][2]['second']['color'], 'blue')
 
+    def test_graph_with_tuple_nodes(self):
+        G = nx.Graph()
+        G.add_edge((0, 0), (1, 0), color=[255, 255, 0])
+        d = node_link_data(G)
+        with NamedTemporaryFile('w', delete=False) as file:
```
Comment (hagberg) [inline networkx/readwrite/json_graph/tests/test_node_link.py]: Let's just use json.dumps and json.loads here to avoid writing a temporary file.

**Comment by hagberg** [comment]
It's good, and needs one small fix in the tests.

File: networkx/readwrite/json_graph/node_link.py [hunk 2]
```diff
@@ -10,7 +10,8 @@
 from itertools import chain, count
 import json
 import networkx as nx
-from networkx.utils import make_str
+from networkx.utils import make_str, to_tuple
+from tempfile import NamedTemporaryFile
```
Comment (jfinkels) [inline networkx/readwrite/json_graph/node_link.py]: I don't think this needs to be here.

File: networkx/utils/misc.py [hunk 3]
```diff
@@ -254,3 +254,19 @@ def groups(many_to_one):
...
+
+        >>> from networkx.utils import to_tuple
+        >>> a_list = [1, 2, [1, 4]]
+        >>> to_tuple(a_list)
+        (1, 2, (1, 4))
+
+    """
+    if isinstance(x, (tuple, list)):
+        return tuple([to_tuple(xx) for xx in x])
+    return x
```
Comment (jfinkels) [inline networkx/utils/misc.py]: Three things: First, there is a missing newline here (and in the `test_misc` file). Second, recursive functions should have a base case first followed by the recursive case:

    if not isinstance(x, (tuple, list)):
        return x
    return tuple(map(to_tuple, x))

Finally, there's no need to create a list as the argument to `tuple`, just a generator expression (or an iterator, as I've done above).

File: networkx/readwrite/json_graph/tests/test_node_link.py (hunk 1 above)
Comment (fjmalass) [inline networkx/readwrite/json_graph/tests/test_node_link.py]: Implemented.. thanks..

File: networkx/utils/misc.py (hunk 3 above)
Comment (fjmalass) [inline networkx/utils/misc.py]: implemented thx.

**Pull request merged**